### PR TITLE
Add source information to GT4Py IRs

### PR DESCRIPTION
**Pull request by tbennun**
## Description

This PR adds and propagates source line information to the GT4Py IRs:

- [x] Python AST -> Definition IR
- [x] Definition IR -> GTIR
- [x] GTIR -> OIR
- [x] OIR -> SDFG -> OIR

### CODE REVIEW

File: src/gtc/dace/dace_to_oir.py [hunk 1]
```diff
@@ -74,7 +74,7 @@ def sdfg_arrays_to_oir_decls(sdfg: dace.SDFG) -> Tuple[List[oir.Decl], List[oir.
     return params, decls
 
 
-def convert(sdfg: dace.SDFG) -> oir.Stencil:
+def convert(sdfg: dace.SDFG, stencil: Optional[oir.Stencil] = None) -> oir.Stencil:
```
Comment (jdahm) [inline src/gtc/dace/dace_to_oir.py:77]: This does not actually use stencil, only the attribute on stencil. Before merging, could you convert this to something like`*, loc: Optional[common.SourceLocation] = None`?

**Comment by jdahm** [comment]
Before we merge I'd like to see if we could implement an idea from #587 to simplify this.

**Comment by havogt** [comment]
@jdahm Are you actively working on the idea of #587?

File: src/gtc/dace/dace_to_oir.py (hunk 1 above)
Comment (tbennun) [reply src/gtc/dace/dace_to_oir.py:77]: sounds good 👍 will do

**Comment by DropD** [comment]
I think #587 would indeed simplify this as it would many other things. However, I also think this is useful in the meantime even if it will be made obsolete relatively soon.

**Review by DropD — APPROVED** [review]
Looks good and I suppose it will help when inspecting SDFGs, I wish I had the source locations there while working on the halo exchanges.

**Comment by tbennun** [comment]
> Looks good and I suppose it will help when inspecting SDFGs, I wish I had the source locations there while working on the halo exchanges.

yes, that is exactly the intent :) it will also help name the GPU kernels after their respective stencil names for clarity in profiles

**Comment by havogt** [comment]
bors try

**Pull request merged**